launcher/modplatform/modrinth/ModrinthPackExportTask.cpp

```diff
@@ -0,0 +1,309 @@
+                QUrl& url = mod->metadata()->url;
+                // most likely some of these may be from curseforge
+                if (!url.isEmpty() && BuildConfig.MODRINTH_MRPACK_HOSTS.contains(url.host())) {
+                    qDebug() << "Resolving" << relative << "from index";
+
+                    // we've already read it
+                    // let's go back!
+                    openFile.seek(openFile.size());
+
+                    QCryptographicHash hash2(QCryptographicHash::Algorithm::Sha1);
+                    hash2.addData(data);
+
+                    ResolvedFile file{ hash2.result().toHex(), hash.result().toHex(), url.toString(), openFile.size() };
+                    resolvedFiles[relative] = file;
+
+                    // nice! we've managed to resolve based on local metadata!
+                    // no need to enqueue it
+                    continue;
+                }
+            }
+        }
+
+        qDebug() << "Enqueueing" << relative << "for Modrinth query";
+        pendingHashes[relative] = hash.result().toHex();
+    }
+
+    makeApiRequest();
+}
+
+void ModrinthPackExportTask::makeApiRequest()
+{
+    if (pendingHashes.isEmpty())
+        buildZip();
+    else {
+        QByteArray* response = new QByteArray;
+        task = api.currentVersions(pendingHashes.values(), "sha512", response);
+        connect(task.get(), &NetJob::succeeded, [this, response]() { parseApiResponse(response); });
+        connect(task.get(), &NetJob::failed, this, &ModrinthPackExportTask::emitFailed);
+        task->start();
+    }
+}
+
+void ModrinthPackExportTask::parseApiResponse(QByteArray* response)
+{
+    task = nullptr;
+
+    try {
+        QJsonDocument doc = Json::requireDocument(*response);
+
+        QMapIterator<QString, QString> iterator(pendingHashes);
+        while (iterator.hasNext()) {
+            iterator.next();
+
+            QJsonObject obj = doc[iterator.value()].toObject();
+            if (obj.isEmpty())
+                continue;
+
+            QJsonArray files = obj["files"].toArray();
+            if (auto fileIter = std::find_if(files.begin(), files.end(),
+                                             [&iterator](const QJsonValue& file) { return file["hashes"]["sha512"] == iterator.value(); });
+                fileIter != files.end()) {
+                // map the file to the url
+                resolvedFiles[iterator.key()] =
+                    ResolvedFile{ fileIter->toObject()["hashes"].toObject()["sha1"].toString(), iterator.value(),
+                                  fileIter->toObject()["url"].toString(), fileIter->toObject()["size"].toInt() };
+            }
+        }
+    } catch (Json::JsonException& e) {
+        qWarning() << "Failed to parse versions response" << e.what();
+    }
+    pendingHashes.clear();
+}
+
+void ModrinthPackExportTask::buildZip()
+{
+    static_cast<void>(QtConcurrent::run(QThreadPool::globalInstance(), [this]() {
```

Comment:
You mentions on the discord that there was a segfault if you aborted too quickly, is that still true?

If so I think this may play a role, I don't think this extra thread is helping here. it's not enabling concurrency an allowing the abort boolean to change mid run. 

I could be wrong on this of course but this may cause a race condition if the thread tries to read that boolean while the GUI thread updates it.

Reply:
Without it the gui is unresponsive. The segmentation fault happens when aborting the Modrinth upload task.

Reply:
I'm happy to take other suggestions on how to make it concurrent!